apps/wolfssh/common.c

```diff
@@ -186,20 +187,179 @@ void ClientIPOverride(int flag)
 #endif /* WOLFSSH_CERTS */
 
 
+static int AppendKeyToFile(const char* filename, const char* name,
+        const char* type, const char* key)
+{
+    FILE *f;
+
+    f = fopen(filename, "a");
+    fprintf(f, "%s %s %s\n", name, type, key);
+    fclose(f);
+
+    return 0;
+}
+
+
+static int FingerprintKey(const byte* pubKey, word32 pubKeySz, char* out)
+{
+    wc_Sha256 sha;
+    byte digest[WC_SHA256_DIGEST_SIZE];
+    char fp[48] = { 0 };
+    word32 fpSz = sizeof(fp);
+
+    wc_InitSha256(&sha);
+    wc_Sha256Update(&sha, pubKey, pubKeySz);
+    wc_Sha256Final(&sha, digest);
+
+    Base64_Encode_NoNl(digest, sizeof(digest), (byte*)fp, &fpSz);
+
+    if (fp[fpSz] == '=') {
+        fp[fpSz] = 0;
+    }
+
+    strcat(out, "SHA256:");
+    strcat(out, fp);
+
+    return 0;
+}
+
+
+static int GetConfirmation(void)
+{
+    int c, confirmed = 0;
+
+    printf("Y/n: ");
+    c = getchar();
+
+    if (c == 'Y') {
+        confirmed = 1;
+    }
+
+    return confirmed;
+}
+
+
 int ClientPublicKeyCheck(const byte* pubKey, word32 pubKeySz, void* ctx)
 {
-    int ret = 0;
+    char *cursor;
+    char *line;
+    const char *targetName = (const char*)ctx;
+    char *name;
+    char *keyType;
+    char *key;
+    char *knownHosts = NULL;
+    char *knownHostsName = NULL;
+    int ret = 0, found = 0, badMatch = 0, otherMatch = 0;
+    word32 sz, lineCount = 0;
+    char encodedKey[1200] = { 0 };
+    char pubKeyType[54] = { 0 };
+    char fp[56] = { 0 };
+
+    {
+        const char *defaultName = "/.ssh/known_hosts";
+        char *env;
+
+        env = getenv("HOME");
+        sz = (word32)(WSTRLEN(env) + WSTRLEN(defaultName) + 1);
+        knownHostsName = (char*)WMALLOC(sz, NULL, 0);
+        if (knownHostsName != NULL) {
+            WSTRCPY(knownHostsName, env);
+            WSTRCAT(knownHostsName, defaultName);
+        }
+    }
 
-    #ifdef DEBUG_WOLFSSH
-        printf("Sample public key check callback\n"
-               "  public key = %p\n"
-               "  public key size = %u\n"
-               "  ctx = %s\n", pubKey, pubKeySz, (const char*)ctx);
-    #else
-        (void)pubKey;
-        (void)pubKeySz;
-        (void)ctx;
-    #endif
+    sz = 0;
+    ret = load_der_file(knownHostsName, (byte**)&knownHosts, &sz);
+
+    /* Get the key type out of the key. */
+    ato32(pubKey, &sz);
+    WMEMCPY(pubKeyType, pubKey + LENGTH_SZ, sz);
+    pubKeyType[sz] = 0;
+
+    sz = (word32)sizeof(encodedKey);
+    ret = Base64_Encode_NoNl(pubKey, pubKeySz, (byte*)encodedKey, &sz);
+    FingerprintKey(pubKey, pubKeySz, fp);
+
+    cursor = knownHosts;
+    while (cursor) {
+        lineCount++;
+        line = WSTRSEP(&cursor, "\n");
```

Comment:
Heap buffer overflow can happen around this area of code, please add additional sanity checks on upper bounds of cursor.